pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -55,268 +136,511 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
-	r.drainer.Force = true
+			r.reqLogger.Infof("holding lease. Maintenance mode ongoing. node: %s\n", nodeName)
+		}
 
-	//Continue even if there are pods using emptyDir (local data that will be deleted when the node is drained).
-	//This is necessary for removing any pod that utilizes an emptyDir volume.
-	//The VirtualMachineInstance Pod does use emptryDir volumes,
-	//however the data in those volumes are ephemeral which means it is safe to delete after termination.
-	r.drainer.DeleteLocalData = true
+	case LeaseStatusActiveOwnedByDifferentHolder:
+		{
+			if statusAnnoData.Status == NodeStateInit {
+				r.reqLogger.Infof("wait to acquire lease. requeue. node: %s\n", nodeName)
+				r.setStatusAnnotation(node, statusAnnoData)
+				return true, &reconcile.Result{Requeue: true, RequeueAfter: 5 * time.Second}, nil
+			}
 
-	//Ignore DaemonSet-managed pods.
-	//This is required because every node running a VirtualMachineInstance will also be running our helper DaemonSet called virt-handler.
-	//This flag indicates that it is safe to proceed with the eviction and to just ignore DaemonSets.
-	r.drainer.IgnoreAllDaemonSets = true
+			statusAnnoData.Status = NodeStateLeaseNotOwnedByNode
+			r.setStatusAnnotation(node, statusAnnoData)
 
-	//Period of time in seconds given to each pod to terminate gracefully. If negative, the default value specified in the pod will be used.
-	r.drainer.GracePeriodSeconds = -1
+			err := fmt.Errorf("lease got acquired by someone while request was active. node: %s", nodeName)
+			return true, &reconcile.Result{}, err
+		}
 
-	// TODO - add logical value or attach from the maintancene CR
-	//The length of time to wait before giving up, zero means infinite
-	r.drainer.Timeout = time.Minute
+	case LeaseStatusFail:
+		{
+			r.reqLogger.Errorf("failed to get lease object. node: %s error: %v", nodeName, err)
 
-	// TODO - consider pod selectors (only for VMIs + others ?)
-	//Label selector to filter pods on the node
-	//r.drainer.PodSelector = "kubevirt.io=virt-launcher"
+			statusAnnoData.Status = NodeStateFailedToGetLease
+			r.setStatusAnnotation(node, statusAnnoData)
 
-	cs, err := kubernetes.NewForConfig(config)
-	if err != nil {
-		return err
+			return true, &reconcile.Result{}, err
+		}
 	}
-	r.drainer.Client = cs
-	r.drainer.DryRun = false
 
-	return nil
+	// TODO:  how to handle possible failure to modify lease time ?
+	r.adjustLeaseTimeIfRequired(specAnnoData, statusAnnoData, lease)
+
+	return false, nil, nil
 }
 
-var _ reconcile.Reconciler = &ReconcileNodeMaintenance{}
+func (r *ReconcileNodeMaintenance) adjustLeaseTimeIfRequired(specAnnoData *ReconcileNodeMaintenanceSpecInfo, statusAnnoData *ReconcileNodeMaintenanceStatusInfo, lease *coordv1beta1.Lease) error {
+	duration := statusAnnoData.OpDeadline.Sub(statusAnnoData.OpStart)
+	durationSeconds := int64(duration.Nanoseconds()) / int64(time.Second)
+
+	diffTime := int64(specAnnoData.RequestedMaintenanceWindowSeconds) - durationSeconds
```

Comment:
Why not just compare lease.Spec.LeaseDurationSeconds with the value in the annotation?